core: Terraform (master) crash in field_reader_config.go

I'm getting a crash related to a null reference being used. I'll try and see what I can do in the terraform source as a bug.

Stacktrace: https://gist.github.com/adamdecaf/51c5ad7344bfc28b363e

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x1571e5]
 
goroutine 121 [running]:
github.com/hashicorp/terraform/helper/schema.(*Set).add(0xc2080aa860, 0x3931c0, 0xc2080a0720, 0xc2080564d0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/set.go:125 +0x95
github.com/hashicorp/terraform/helper/schema.(*ConfigFieldReader).readSet(0xc2080aa620, 0xc2080722b0, 0x1, 0x1, 0xc2080564d0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/field_reader_config.go:202 +0x5ed
github.com/hashicorp/terraform/helper/schema.(*ConfigFieldReader).readField(0xc2080aa620, 0xc2080722b0, 0x1, 0x1, 0x737300, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/field_reader_config.go:96 +0x619
github.com/hashicorp/terraform/helper/schema.(*ConfigFieldReader).ReadField(0xc2080aa620, 0xc2080722b0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/field_reader_config.go:25 +0x86
github.com/hashicorp/terraform/helper/schema.(*MultiLevelFieldReader).ReadFieldExact(0xc2080aa660, 0xc2080722b0, 0x1, 0x1, 0x4dc750, 0x6, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/field_reader_multi.go:31 +0x247
github.com/hashicorp/terraform/helper/schema.(*ResourceData).get(0xc2080b4fc0, 0xc2080722b0, 0x1, 0x1, 0xc208072202, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/resource_data.go:387 +0x2a2
github.com/hashicorp/terraform/helper/schema.(*ResourceData).getChange(0xc2080b4fc0, 0x52cb30, 0x17, 0x1b1201, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/resource_data.go:356 +0x1d0
github.com/hashicorp/terraform/helper/schema.(*ResourceData).diffChange(0xc2080b4fc0, 0x52cb30, 0x17, 0x0, 0x0, 0x0, 0x0, 0x1ba5ec)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/resource_data.go:333 +0x87
github.com/hashicorp/terraform/helper/schema.schemaMap.diffSet(0xc20803d350, 0x52cb30, 0x17, 0xc2080564d0, 0xc2080b1ec0, 0xc2080b4fc0, 0xc20801e800, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/schema.go:707 +0x84
github.com/hashicorp/terraform/helper/schema.schemaMap.diff(0xc20803d350, 0x52cb30, 0x17, 0xc2080564d0, 0xc2080b1ec0, 0xc2080b4fc0, 0xc2080aa500, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/schema.go:493 +0x2f6
github.com/hashicorp/terraform/helper/schema.schemaMap.Diff(0xc20803d350, 0xc2080a0570, 0xc2080a0480, 0xc2081135b8, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/schema.go:257 +0x1fb
github.com/hashicorp/terraform/helper/schema.(*Resource).Diff(0xc20803ba00, 0xc2080a0570, 0xc2080a0480, 0x11, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/resource.go:162 +0x51
github.com/hashicorp/terraform/helper/schema.(*Provider).Diff(0xc20803d380, 0xc2080a2c80, 0xc2080a0570, 0xc2080a0480, 0x0, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/helper/schema/provider.go:167 +0x1b9
github.com/hashicorp/terraform/terraform.(*EvalDiff).Eval(0xc2080775c0, 0x913088, 0xc208070240, 0x0, 0x0, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/terraform/eval_diff.go:86 +0x25a
github.com/hashicorp/terraform/terraform.EvalRaw(0x913900, 0xc2080775c0, 0x913088, 0xc208070240, 0x0, 0x0, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x254
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc20800ba00, 0x913088, 0xc208070240, 0x0, 0x0, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:10 +0xf9
github.com/hashicorp/terraform/terraform.EvalRaw(0x9131c0, 0xc20800ba00, 0x913088, 0xc208070240, 0x0, 0x0, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x254
github.com/hashicorp/terraform/terraform.(*EvalOpFilter).Eval(0xc208077680, 0x913088, 0xc208070240, 0x0, 0x0, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/terraform/eval_filter_operation.go:37 +0x71
github.com/hashicorp/terraform/terraform.EvalRaw(0x9131e8, 0xc208077680, 0x913088, 0xc208070240, 0x0, 0x0, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x254
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc20800b840, 0x913088, 0xc208070240, 0x0, 0x0, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:10 +0xf9
github.com/hashicorp/terraform/terraform.EvalRaw(0x9131c0, 0xc20800b840, 0x913088, 0xc208070240, 0x0, 0x0, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x254
github.com/hashicorp/terraform/terraform.Eval(0x9131c0, 0xc20800b840, 0x913088, 0xc208070240, 0x0, 0x0, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/terraform/eval.go:34 +0x79
github.com/hashicorp/terraform/terraform.func·027(0x4a4440, 0xc2080b0e90, 0x0, 0x0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/terraform/graph.go:158 +0x754
github.com/hashicorp/terraform/dag.func·004(0x4a4440, 0xc2080b0e90, 0xc2080b4660, 0xc2080b46c0)
	/Users/adam/src/go/src/github.com/hashicorp/terraform/dag/dag.go:187 +0x106
created by github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk
	/Users/adam/src/go/src/github.com/hashicorp/terraform/dag/dag.go:196 +0x8ef
```

--- a/helper/schema/provider.go
+++ b/helper/schema/provider.go
@@ -97,6 +97,13 @@ func (p *Provider) Input(
 
 // Validate implementation of terraform.ResourceProvider interface.
 func (p *Provider) Validate(c *terraform.ResourceConfig) ([]string, []error) {
+	if err := p.InternalValidate(); err != nil {
+		return nil, []error{fmt.Errorf(
+			"Internal validation of the provider failed! This is always a bug\n"+
+				"with the provider itself, and not a user issue. Please report\n"+
+				"this bug:\n\n%s", err)}
+	}
+
 	return schemaMap(p.Schema).Validate(c)
 }
 

--- a/helper/schema/provider_test.go
+++ b/helper/schema/provider_test.go
@@ -117,6 +117,36 @@ func TestProviderResources(t *testing.T) {
 	}
 }
 
+func TestProviderValidate(t *testing.T) {
+	cases := []struct {
+		P      *Provider
+		Config map[string]interface{}
+		Err    bool
+	}{
+		{
+			P: &Provider{
+				Schema: map[string]*Schema{
+					"foo": &Schema{},
+				},
+			},
+			Config: nil,
+			Err:    true,
+		},
+	}
+
+	for i, tc := range cases {
+		c, err := config.NewRawConfig(tc.Config)
+		if err != nil {
+			t.Fatalf("err: %s", err)
+		}
+
+		_, es := tc.P.Validate(terraform.NewResourceConfig(c))
+		if (len(es) > 0) != tc.Err {
+			t.Fatalf("%d: %#v", i, es)
+		}
+	}
+}
+
 func TestProviderValidateResource(t *testing.T) {
 	cases := []struct {
 		P      *Provider

